runtime/eventlog.c

```diff
@@ -0,0 +1,257 @@
+struct event_buffer {
+  uintnat ev_generated;
+  struct event events[EVENT_BUF_SIZE];
+};
+
+static struct event_buffer* evbuf;
+
+void setup_evbuf()
+{
+  CAMLassert(!evbuf);
+  evbuf = malloc(sizeof(*evbuf));
+
+  if (!evbuf) return;
+
+  evbuf->ev_generated = 0;
+}
+
+void setup_eventlog_file()
+{
+  char *filename;
+  char *ocaml_eventlog_filename;
+
+  ocaml_eventlog_filename = caml_secure_getenv("OCAML_EVENTLOG_FILE");
+  if (ocaml_eventlog_filename) {
+    filename = malloc(strlen(ocaml_eventlog_filename) + 64);
+    sprintf(filename, "%s.%d.eventlog",
+            ocaml_eventlog_filename, eventlog_startup_pid);
+  } else {
+    filename = malloc(64);
+    sprintf(filename, "caml-eventlog-%d", eventlog_startup_pid);
+  }
+
+  output = fopen(filename, "wb");
+  if (output) {
+    fwrite(&header, sizeof(struct ctf_stream_header), 1, output);
+    fflush(output);
+  } else {
+    caml_eventlog_enabled = 0;
+  }
+  free(filename);
+}
+
+static void flush_events(FILE* out, struct event_buffer* eb)
+{
+  uintnat i;
+  uintnat n = eb->ev_generated;
+
+  struct ctf_event_header ev_flush;
+  ev_flush.id = EV_FLUSH;
+  ev_flush.timestamp = caml_time_counter() - eventlog_startup_timestamp;
+  ev_flush.pid = eventlog_startup_pid;
+
+  for (i = 0; i < n; i++) {
+    struct event ev = eb->events[i];
+    ev.header.pid = eventlog_startup_pid;
+
+    fwrite(&ev.header, sizeof(struct ctf_event_header), 1, out);
+
+    switch (ev.header.id)
+    {
+    case EV_ENTRY:
+      fwrite(&ev.phase, sizeof(uint8_t), 1, out);
+      break;
+    case EV_EXIT:
+      fwrite(&ev.phase, sizeof(uint8_t), 1, out);
+      break;
+    case EV_COUNTER:
+      fwrite(&ev.count, sizeof(uint32_t), 1, out);
+      fwrite(&ev.counter_kind, sizeof(uint8_t), 1, out);
+      break;
+    case EV_ALLOC:
+      fwrite(&ev.count, sizeof(uint32_t), 1, out);
+      fwrite(&ev.alloc_bucket, sizeof(uint8_t), 1, out);
+      break;
+    default:
+      break;
+    }
+  }
+
+  uint64_t flush_duration =
+    (caml_time_counter() - eventlog_startup_timestamp) - ev_flush.timestamp;
+
+  fwrite(&ev_flush, sizeof(struct ctf_event_header), 1, out);
+  fwrite(&flush_duration, sizeof(uint64_t), 1, out);
+}
+
+static void teardown_eventlog()
+{
+  if (evbuf) {
+    flush_events(output, evbuf);
+    free(evbuf);
+    evbuf = 0;
+  }
+  if (output) {
+    fclose(output);
+    output = 0;
+  }
+}
+
+void caml_setup_eventlog()
+{
+  if (caml_secure_getenv("OCAML_EVENTLOG_ENABLED"))
+    caml_eventlog_enabled = 1;
+  if (!caml_eventlog_enabled) return;
+
+  eventlog_startup_timestamp = caml_time_counter();
+  eventlog_startup_pid = getpid();
+  setup_eventlog_file();
+
+  atexit(&teardown_eventlog);
+}
+
+static void post_event(ev_gc_phase phase, ev_gc_counter counter_kind,
+                       uint8_t bucket, uint32_t count, ev_type ty)
+{
+  uintnat i;
+  struct event* ev;
+
+  if (!caml_eventlog_enabled) return;
+  if (eventlog_paused) return;
+  if (!evbuf) setup_evbuf();
```

Comment:
This is where you assume that `setup_evbuf` always succeeds in initialising `evbuf` to non-NULL.

Reply:
it is now handled, failure to allocate event buffer results in a `caml_fatal_error` call

Reply:
I would like to avoid calling `caml_fatal_error` (even if only theoretically) wherever it is possible to raise an exception instead. You are allocating at the first occurring event, where you do not know whether you can raise. I see two solutions to avoid this:
- you could call `setup_evbuf` at the moment eventlog is enabled. For now it is in `caml_gc_init` so it is still necessary to call `caml_fatal_error`, but I find it more sensible to fail there than during execution. But if later you wanted to allow enabling eventlog at runtime, you could raise then (out of the C stub).
- if you really want to delay its allocation, you could delay the call to `setup_evbuf` to the first time eventlog is resumed, whenever it starts paused. Then it is safe to raise at that point.

Reply:
I don't think delaying allocation is an interesting property to keep and it make state a bit too fuzzy to track.
I will go with the first idea, whether it is enabled or paused on startup, allocating if eventlog was explicitly enabled seems fair enough.